cassiopeia/solver/ilp_solver_utilities.pyx

```diff
@@ -93,13 +231,90 @@ def infer_layer_of_potential_graph(
         for k in range(len(top_ancestors)):
             if distance_to_ancestors[k] == min_distance_to_ancestor:
 
-                parent = top_ancestors[k][:dim]
+                # The first (2*dim-1) characters correspond to the parent
+                parent = "|".join(top_ancestors[k].split("|")[:dim])
                 edge2 = top_ancestors[k]
-                edge1 = np.concatenate((parent, sample1))
+                edge1 = parent + "|" + sample1
+
+                new_edges.add(edge1)
+                new_edges.add(edge2)
+
+                layer.add(parent)
+
+    return np.array(list(layer)), np.array(list(new_edges))
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def get_lca_characters_cython(
+    arr1: np.array(),
+    arr2: np.array(),
+    n_char: int,
+    missing_state_indicator: str,
+) -> str:
+    """Cython-optimized LCA inference.
+
+    A utility function for the inference of the latest common ancestor between
+    two samples. Here, both samples are lists of strings where each value
+    corresponds to a state. The procedure infers an ancestral state according to
+    Camin-Sokal parsimony and returns an ancestor character string where each
+    state is delimited by `|`.
+
+    Args:
+        arr1: Character array of sample 1
+        arr2: Character array of sample 2
+        n_char: Number of characters in array
+        missing_state_indicator: Indicator for missing data.
+
+    Returns:
+        The character string of the inferred ancestor, where each character is
+            delimited by a `|`.
+    """
+
+    cdef list ancestor = ["0"]*n_char
+    cdef int i
+
+    for i in range(n_char):
+
+        if arr1[i] == arr2[i]:
+            ancestor[i] = arr1[i]
+
+        else:
+            if arr1[i] == missing_state_indicator:
+                ancestor[i] = arr2[i]
+            elif arr2[i] == missing_state_indicator:
+                ancestor[i] = arr1[i]
+
+    return "|".join(ancestor)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def simple_hamming_distance_cython(
```

Comment:
Given that this is ignoring missing data should we add this in the function name?

Reply:
I think it's fine -- it shouldn't be exposed to users and I believe this information is in the docstring.